scheduler/replica_scheduler.go

```diff
@@ -224,33 +223,31 @@ func (rcs *ReplicaScheduler) getDiskCandidates(nodeInfo map[string]*longhorn.Nod
 		}
 		if _, ok := usedNodes[nodeName]; !ok {
 			unusedNodes[nodeName] = node
-			if _, ok := usedZones[node.Status.Zone]; !ok {
-				unusedNodesInNewZones[nodeName] = node
-			}
 		}
 		if _, ok := usedZones[node.Status.Zone]; !ok {
-			nodesInUnusedZones[nodeName] = node
+			unusedNodesInUnusedZones[nodeName] = node
 		}
 	}
 
+	// In all cases, we should try to use a disk on an unused node in an unused zone first. Don't bother considering
+	// zoneSoftAntiAffinity and nodeSoftAntiAffinity settings if such disks are available.
+	diskCandidates, errors := getDiskCandidatesFromNodes(unusedNodesInUnusedZones)
+	if len(diskCandidates) > 0 {
+		return diskCandidates, nil
+	}
+	multiError.Append(errors)
+
 	switch {
 	case !zoneSoftAntiAffinity && !nodeSoftAntiAffinity:
-		diskCandidates, errors := getDiskCandidatesFromNodes(unusedNodesInNewZones)
-		if len(diskCandidates) > 0 {
-			return diskCandidates, nil
-		}
-		multiError.Append(errors)
+		fallthrough
+	// Same as the above. If we cannot schedule two replicas in the same zone, we cannot schedule them on the same node.
+	case !zoneSoftAntiAffinity && nodeSoftAntiAffinity:
```

Comment:
Sorry... I don't get how you guarantee `!zoneSoftAntiAffinity` with the below code...

Reply:
Interesting. This is not a behavioral change. We previously used the same code in both blocks that had `!zoneSoftAntiAffinity`. I am guessing the thinking was that we are just replacing an evicting replica and the replica already was scheduled to this node, so rescheduling it to this node is fine.

The only way I can think of this being a problem is if the zone a node is in has changed since we originally scheduled the evicting replica. I think this is possible (maybe the user redefined zones or originally did not define zones), so we probably need to make changes.

This might have made sense to handle in a separate ticket, but since we are already refactoring the scheduler in this one, I'll see what I can do.

- [ ] Ensure we only replace an evicting replica if doing so would still respect `!zoneSoftAntiAffinity`.

Reply:
> I am guessing the thinking was that we are just replacing an evicting replica and the replica already was scheduled to this node, so rescheduling it to this node is fine.

OK. This is why the function has `getDiskCandidatesFromNodes(filterNodesWithLessThanTwoReplicas(nodesWithEvictingReplicas))`. 
But I don't think it always respects the affinity rules as the affinities may change before the eviction happens. In other words, a node containing an existing/evicting replica may no longer match the rules. Probably we need to re-check rather than blindly use the nodes containing one evicting replica.

Besides, with the modification you make for case `!zoneSoftAntiAffinity && nodeSoftAntiAffinity`, there is no way to schedule a replica to a **used** node in unused zones.

Reply:
> Besides, with the modification you make for case !zoneSoftAntiAffinity && nodeSoftAntiAffinity, there is no way to schedule a replica to a used node in unused zones.

This was the discussion that lead to the simplification in the first place (https://github.com/longhorn/longhorn-manager/pull/2094#pullrequestreview-1555771391 and internal). What is a "used node in an unused zone"? If a node is used, then the zone it is in is also used. Our criteria for determining whether a zone is used is whether it contains a used node.

I tried a couple of times to diagram this, but struggled to produce something I liked. Here's one attempt at it:

![image](https://github.com/longhorn/longhorn-manager/assets/24213029/ce608dd9-4a0f-4e31-b5e5-2017db2ea2b5)

Is there is a counterexample to this I'm missing?

Reply:
> But I don't think it always respects the affinity rules as the affinities may change before the eviction happens. In other words, a node containing an existing/evicting replica may no longer match the rules. Probably we need to re-check rather than blindly use the nodes containing one evicting replica.

This is also something interesting to think about, though I would argue for doing it in a separate ticket if needed. While this PR introduces minor changes to scheduling behavior, it is mostly just a logic refactor. We already are scheduling replicas to nodes that are evicting them, and reconsidering this makes it difficult to focus on the purpose of this PR (disk anti-affinity).

To kick off that discussion though, can you clarify this with an example? When you say affinity, or you referring to node, zone, and disk anti-affinities? These are already covered. For example in the !zoneSoftAntiAffinity && !nodeSoftAntiAffinity case, we know we cannot schedule to a used node or used zone. We try the obvious (schedule to an unused node in an unused zone) first, then we schedule to a node which:

- Is in a zone where all replicas are evicting (this means soon the zone will be unused).
- Has only one evicting replica (this means soon the node will have no replica).

I am not sure what else we need to check regarding affinities.

If you are referring to other scheduling constraints (e.g., is a disk actually schedulable, are the tags appropriate, etc.), we handle that after this logic in `filterNodeDisksForReplica`.

Reply:
> If a node is used, then the zone it is in is also used. Our criteria for determining whether a zone is used is whether it contains a used node.

Got it. Great explanation!

> To kick off that discussion though, can you clarify this with an example?

Sure. Here is an example: Let's say there are 3 nodes in the cluster. node1 and node2 belongs to zoneA while node3 belongs to zoneB
1. When `zoneSoftAntiAffinity && !nodeSoftAntiAffinity`. We create a 3-replica volume. Each replica is on a separate replica
2. Later on we update the affinity rule to `!zoneSoftAntiAffinity && !nodeSoftAntiAffinity`
3. Somehow we evict the replica on node1.
4. Ideally, the volume cannot do scheduling for the new replica. But here `nodesWithEvictingReplicas()` returns node1 and `filterNodesWithLessThanTwoReplicas(filterNodesInOnlyEvictingZones())` cannot filter it out. As a result, the new replica will be scheduled to node1, which breaks the new rule. Is this expected?

Reply:
I went ahead and implemented a new test similar to your example. The final test as described below only uses one zone, but I also ran it with two zones as described.

1. Create three nodes in the same zone. Each node has a replica, but the replica on node1 is evicting.
2. Try to schedule another replica.
3. Verify that replica scheduling fails.

The test passes.

In your exact example, we fail to schedule a replica because `filterNodesInOnlyEvictingZones` actually does filter out `node1`. `zoneA` is NOT an only evicting zone, so we are not allowed to schedule to it.

![image](https://github.com/longhorn/longhorn-manager/assets/24213029/0508f703-860d-478e-9281-f64d2a9ff0ed)

Reply:
While looking into the example and writing the new test case, I further simplified/clarified the scheduling for the evicting replica case. I do not see any behavioral changes, but I think it makes the code a bit easier to read.